File: app/chrome-ext/src/lib/watermark-engine.ts

export enum WatermarkSize {
    Small,
    Large,
}

interface WatermarkPosition {
    marginRight: number;
    marginBottom: number;
}

interface AlphaMap {
    width: number;
    height: number;
    data: Float32Array;
}

export class WatermarkEngine {
    private alphaMapSmall: AlphaMap | null = null;
    private alphaMapLarge: AlphaMap | null = null;
    private logoValue: number = 255.0;

    constructor() { }

    public async init() {
        // Load assets from public/assets folder
        const bgSmall = await this.loadImage('assets/bg_48.png');
        const bgLarge = await this.loadImage('assets/bg_96.png');

        this.alphaMapSmall = this.calculateAlphaMap(bgSmall);
        this.alphaMapLarge = this.calculateAlphaMap(bgLarge);
    }

    private loadImage(src: string): Promise<HTMLImageElement> {
        return new Promise((resolve, reject) => {
            const img = new Image();
            img.onload = () => resolve(img);
            img.onerror = reject;
            img.src = src;
        });
    }

    private calculateAlphaMap(img: HTMLImageElement): AlphaMap {
        const canvas = document.createElement('canvas');
        canvas.width = img.width;
        canvas.height = img.height;
        const ctx = canvas.getContext('2d');
        if (!ctx) throw new Error("Could not get canvas context");

        ctx.drawImage(img, 0, 0);
        const imageData = ctx.getImageData(0, 0, img.width, img.height);
        const data = imageData.data; // RGBA

        const alphaData = new Float32Array(img.width * img.height);

        for (let i = 0; i < alphaData.length; i++) {
            const r = data[i * 4];
            const g = data[i * 4 + 1];
            const b = data[i * 4 + 2];

            // max(r, g, b) / 255.0
            const maxVal = Math.max(r, Math.max(g, b));
            alphaData[i] = maxVal / 255.0;
        }

        return {
            width: img.width,
            height: img.height,
            data: alphaData
        };
    }

    public async removeWatermark(imageSource: string | File): Promise<string> {
        // Load target image
        let img: HTMLImageElement;
        if (imageSource instanceof File) {
            const url = URL.createObjectURL(imageSource);
            img = await this.loadImage(url);
            URL.revokeObjectURL(url); // Clean up
        } else {
            img = await this.loadImage(imageSource);
        }

        const width = img.width;
        const height = img.height;

        // Determine size
        let size = WatermarkSize.Small;
        if (width > 1024 && height > 1024) {
            size = WatermarkSize.Large;
        }

        let config: WatermarkPosition;
        let alphaMap: AlphaMap | null;

        if (size === WatermarkSize.Small) {
            config = { marginRight: 32, marginBottom: 32 };
            alphaMap = this.alphaMapSmall;
        } else {
            config = { marginRight: 64, marginBottom: 64 };
            alphaMap = this.alphaMapLarge;
        }

        if (!alphaMap) throw new Error("Assets not initialized");

        // Calculate position
        const logoW = alphaMap.width;
        const logoH = alphaMap.height;

        let x = width - config.marginRight - logoW;
        let y = height - config.marginBottom - logoH;

        // Saturating sub equivalent
        if (x < 0) x = 0;
        if (y < 0) y = 0;

        // Process
        const canvas = document.createElement('canvas');
        canvas.width = width;
        canvas.height = height;
        const ctx = canvas.getContext('2d');
        if (!ctx) throw new Error("Canvas context missing");

        ctx.drawImage(img, 0, 0);
        const imageData = ctx.getImageData(0, 0, width, height);

        this.applyRemoval(imageData, alphaMap, x, y);

        ctx.putImageData(imageData, 0, 0);

        return canvas.toDataURL('image/png');
    }

    private applyRemoval(image: ImageData, alphaMap: AlphaMap, x: number, y: number) {
        const imgW = image.width;
        const imgH = image.height;
        const mapW = alphaMap.width;
        const mapH = alphaMap.height;

        const xEnd = Math.min(x + mapW, imgW);
        const yEnd = Math.min(y + mapH, imgH);

        const alphaThreshold = 0.002;
        const maxAlpha = 0.99;

        const data = image.data;

        for (let currY = y; currY < yEnd; currY++) {
            const mapY = currY - y;
            for (let currX = x; currX < xEnd; currX++) {
                const mapX = currX - x;

                const alpha = alphaMap.data[mapY * mapW + mapX];

                if (alpha < alphaThreshold) continue;

                const effectiveAlpha = Math.min(alpha, maxAlpha);
                const oneMinusAlpha = 1.0 - effectiveAlpha;

                const idx = (currY * imgW + currX) * 4;

                // RGB channels
                for (let c = 0; c < 3; c++) {
                    const val = data[idx + c];
                    // original = (val - alpha * logo_value) / one_minus_alpha
                    let original = (val - effectiveAlpha * this.logoValue) / oneMinusAlpha;

                    // clamp
                    if (original < 0) original = 0;
                    if (original > 255) original = 255;

                    data[idx + c] = original;
                }
                // Alpha channel (idx+3) remains 255 usually, or whatever it was
            }
        }
    }
}

File: app/src-tauri/src/watermark_engine.rs
use crate::embedded_assets;
use image::{DynamicImage, GenericImage, GenericImageView, Pixel, Rgba};

#[derive(Clone, Copy, PartialEq)]
pub enum WatermarkSize {
    Small,
    Large,
}

struct WatermarkPosition {
    margin_right: u32,
    margin_bottom: u32,
}

impl WatermarkPosition {
    fn get_position(&self, width: u32, height: u32, logo_w: u32, logo_h: u32) -> (u32, u32) {
        let x = width
            .saturating_sub(self.margin_right)
            .saturating_sub(logo_w);
        let y = height
            .saturating_sub(self.margin_bottom)
            .saturating_sub(logo_h);
        (x, y)
    }
}

pub struct AlphaMap {
    pub width: u32,
    pub height: u32,
    pub data: Vec<f32>,
}

pub struct WatermarkEngine {
    alpha_map_small: AlphaMap,
    alpha_map_large: AlphaMap,
    logo_value: f32,
}

impl WatermarkEngine {
    pub fn new() -> Self {
        // Load potentially resizing if needed, but we assume assets are correct size as per C++
        let bg_small =
            image::load_from_memory(embedded_assets::BG_48_PNG).expect("Failed to load BG_48");
        let bg_large =
            image::load_from_memory(embedded_assets::BG_96_PNG).expect("Failed to load BG_96");

        // Resize if strictly matching C++ paranoia, but assets are static.

        Self {
            alpha_map_small: calculate_alpha_map(&bg_small),
            alpha_map_large: calculate_alpha_map(&bg_large),
            logo_value: 255.0,
        }
    }

    pub fn remove_watermark(&self, image: &mut DynamicImage) {
        let (w, h) = image.dimensions();

        // Determine size
        let size = if w > 1024 && h > 1024 {
            WatermarkSize::Large
        } else {
            WatermarkSize::Small
        };

        let (config, alpha_map) = match size {
            WatermarkSize::Small => (
                WatermarkPosition {
                    margin_right: 32,
                    margin_bottom: 32,
                },
                &self.alpha_map_small,
            ),
            WatermarkSize::Large => (
                WatermarkPosition {
                    margin_right: 64,
                    margin_bottom: 64,
                },
                &self.alpha_map_large,
            ),
        };

        let (pos_x, pos_y) = config.get_position(w, h, alpha_map.width, alpha_map.height);

        self.apply_removal(image, alpha_map, pos_x, pos_y);
    }

    fn apply_removal(&self, image: &mut DynamicImage, alpha_map: &AlphaMap, x: u32, y: u32) {
        let (img_w, img_h) = image.dimensions();
        let map_w = alpha_map.width;
        let map_h = alpha_map.height;

        let x_end = (x + map_w).min(img_w);
        let y_end = (y + map_h).min(img_h);

        let alpha_threshold = 0.002;
        let max_alpha = 0.99;

        for curr_y in y..y_end {
            let map_y = curr_y - y;
            for curr_x in x..x_end {
                let map_x = curr_x - x;

                let alpha = alpha_map.data[(map_y * map_w + map_x) as usize];

                if alpha < alpha_threshold {
                    continue;
                }

                let alpha = alpha.min(max_alpha);
                let one_minus_alpha = 1.0 - alpha;

                let mut pixel = image.get_pixel(curr_x, curr_y);
                // process RGB channels (0, 1, 2)
                for c in 0..3 {
                    let val = pixel[c] as f32;
                    let original = (val - alpha * self.logo_value) / one_minus_alpha;
                    pixel[c] = original.clamp(0.0, 255.0) as u8;
                }
                image.put_pixel(curr_x, curr_y, pixel);
            }
        }
    }
}

fn calculate_alpha_map(img: &DynamicImage) -> AlphaMap {
    let (w, h) = img.dimensions();
    let mut data = Vec::with_capacity((w * h) as usize);

    for y in 0..h {
        for x in 0..w {
            let p = img.get_pixel(x, y);
            // RGB max
            let max_val = p[0].max(p[1]).max(p[2]);
            data.push(max_val as f32 / 255.0);
        }
    }

    AlphaMap {
        width: w,
        height: h,
        data,
    }
}
